File: kotlinbase/src/main/java/com/suifeng/kotlin/base/widget/other/MarqueTextView.java
package com.suifeng.kotlin.base.widget.other;

import android.content.Context;
import android.graphics.Canvas;
import android.support.v7.widget.AppCompatTextView;
import android.text.TextPaint;
import android.util.AttributeSet;

/**
 * 滚动文字
 */
public class MarqueTextView extends AppCompatTextView {


    private StringBuilder mSb = new StringBuilder();
    public boolean isInit = true;

    public MarqueTextView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    public MarqueTextView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public MarqueTextView(Context context) {
        super(context);
    }

    @Override
    public boolean isFocused() {
        return true;
    }

    @Override
    public void onWindowFocusChanged(boolean hasWindowFocus) {

    }

    @Override
    protected void onDraw(Canvas canvas) {
        final StringBuilder sb = mSb;
        if (isInit) {
            int length = sb.length();
            if (length > 0) {
                sb.delete(0, length);
            }
            TextPaint paint = getPaint();
            sb.append(getText());
            float textLength = paint.measureText(sb.toString());
            float needWidth = getWidth() - textLength;
            if (needWidth >= 0) {
                float spaceWidth = paint.measureText(" ");
                int number = (int) (needWidth / spaceWidth + 0.5f) + 2;
                for (int i = 0; i < number; i++) {
                    sb.append(" ");
                }
                setText(sb.toString());
            }
            isInit = false;
        }
        super.onDraw(canvas);
    }
}

File: kotlinbase/src/main/java/com/suifeng/kotlin/base/widget/other/MarqueTextView.kt
package com.suifeng.kotlin.base.widget.other

import android.content.Context
import android.graphics.Canvas
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatTextView

/**
 * 滚动文字
 */
class MarqueTextView : AppCompatTextView {


    private val mSb = StringBuilder()
    var isInit = true

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    constructor(context: Context) : super(context)

    override fun isFocused(): Boolean {
        return true
    }

    override fun onWindowFocusChanged(hasWindowFocus: Boolean) {

    }

    override fun onDraw(canvas: Canvas) {
        val sb = mSb
        if (isInit) {
            val length = sb.length
            if (length > 0) {
                sb.delete(0, length)
            }
            val paint = paint
            sb.append(text)
            val textLength = paint.measureText(sb.toString())
            val needWidth = width - textLength
            if (needWidth >= 0) {
                val spaceWidth = paint.measureText(" ")
                val number = (needWidth / spaceWidth + 0.5f).toInt() + 2
                for (i in 0 until number) {
                    sb.append(" ")
                }
                text = sb.toString()
            }
            isInit = false
        }
        super.onDraw(canvas)
    }
}